### ISSUE TITLE
Run SQLite's VACUUM in a periodic task

### ISSUE THREAD

**Issue by LNSD** [waku-org/nwaku#1154]
### Requirements

Current Status requirements are: 30 days message retention time.

### Context/Rationale

Previous to PR #1025, the VACCUM action was performed every time a message was inserted. The #1025 PR removed it, and it is no longer performed by the node.

Doing a manual test in a copy of nwaku's node SQLite DB the size of the file reduced from 12.GB to 7.5GB (a 40% reduction in size). This was with a DB containing 250.000 messages. At the moment of writing, `status.test` fleet number of stored messages is around 2.4 million messages and increasing.

In addition, a vacuumed store will provide a small improvement in performance in query, insertion, and retention policy execution.

### Acceptance criteria

SQLite vacuuming is a costly task, and it has to be performed periodically to keep the database's performance.

Intuitively, the idea is to run SQLite's VACUUM query:
- [x] On store load, after executing the retention policy/policies.
- [ ] Periodically. For example, every 12, 24, or 48 hours.

**Comment by LNSD** [comment]
Running the VACUUM command in a fixed period can be problematic. When the vacuuming task executes, it will load the CPU. If the load is already high (e.g., while processing requests or relaying messages), this will have an impact on the node's stability/performance.

Currently, it is a trade-off between having a short-term solution to unblock status and having a dynamic task that takes into account the node's load and performs the vacuuming during the low CPU periods.

**Comment by LNSD** [comment]
Copying here @Menduist [comment](https://github.com/status-im/nwaku/issues/1024#issuecomment-1246643582):

> Already mentioned this on discord, but VACUUM only make sense if your db has shrunk significantly
> sqlite will not free unused space, but it will reuse it for future data, so if your dataset has a stable / always growing size, vacuum is almost useless
> 
> In waku's case, it means it's useful in two cases:
> 
> - If you reduce the retention time, you should VACUUM to reclaim the unused space (that can be done at startup)
> - If there was a traffic spike 1 month ago (or whatever the retention time is) that you just got rid of. That's the only case where a timer make sense
>
> Also, before VACUUMing, you can check if it will be useful by checking the ratio between freelist_count and page_count [pragmas](https://www.sqlite.org/pragma.html)
> 
> (all of what I just said completely ignores fragmentation, which is another issue that is solved by VACUUMing)

**Comment by LNSD** [comment]
Extending what @Menduist was says, ther eis this nice read about VACUUMing in SQLite: https://blogs.gnome.org/jnelson/2015/01/06/sqlite-vacuum-and-auto_vacuum/

> ## Heuristic #1: Disk space
> If you’re interested in conserving disk space, the database has to have free pages to release back to the filesystem; it’s as simple as that.  One heuristic is to examine the free versus in-use page counts and trigger a vacuum when a certain threshold is reached (say, 25% of the total pages are free).  That information is available via SQLite’s [page_count](https://www.sqlite.org/pragma.html#pragma_page_count) and [freelist_count](https://www.sqlite.org/pragma.html#pragma_freelist_count) PRAGMAs.
> 
> Another possibility is to multiply the free pages by the [page_size](https://www.sqlite.org/pragma.html#pragma_page_size) PRAGMA and trigger when an absolute size is reached.  This is smarter than vacuuming at a certain ratio, since 25% of 1MB on a modern desktop machine does not represent significant savings in an era of terabyte hard drives.

**Comment by LNSD** [comment]
In the copy of the database I have (with 255k messages), these are the page stats:

```
INF 2022-09-14 17:49:00.284+02:00 sqlite database page stats                 topics="wakunode.setup" tid=2804647 file=sqlite_store.nim:81 pageSize=4096 pages=3134095 unusedPages=1120891
```

Before vacuuming, the database file size was:
```bash
$ ls -la build/store.sqlite3
-rw-r--r--  1 lorenzo lorenzo 12837253120 Sep 14 17:51 store.sqlite3
```

After vacuuming, the database file size is:
```bash
$ ls -la build/store.sqlite3
-rw-r--r-- 1 lorenzo lorenzo 8167907328 Sep 14 17:51 build/store.sqlite3
```
This is a database file size reduction of 36.373402848%

**Comment by jakubgs** [comment]
The main thing to note is that SQLite by default just reuses available free space:

>When content is deleted from an SQLite database, the content is not usually erased but rather the space used to hold the content is marked as being available for reuse.

https://www.sqlite.org/lang_vacuum.html

And fragmentation can be a reason to do it, but it probably doesn't ahve to happen that often:

>Once a sufficient number of rows are removed from a page, and the free space on that page gets to be a substantial fraction of the total space for the page, then the page is merged with adjacent pages, freeing up a whole page for reuse. But as doing this reorganization is expensive, it is deferred until a lot of free space accumulates on the page. (The exact thresholds for when a rebalance occurs are written down some place, but they do not come immediately to my mind, as the whole mechanism just works and we haven't touched it in about 15 years.)

https://stackoverflow.com/questions/59705038/how-to-defragment-consolidate-partially-filled-sqlite-pages-in-order-to-reclaim

**Comment by jakubgs** [comment]
I don't mind having the __option__ to run vacuuming at startup, but I'd disable it by default on the fleets.

**Comment by LNSD** [comment]
Take the PR above (#1156) as a proposal to do some tests and gather some metrics.

**Comment by LNSD** [comment]
With PR #1156 merged and seeing that in the short-mid term having a periodical VACUUM task offers little value, I proceed to put the current issue on hold (a.k.a. move to the icebox) until the priority for implementing this feature is increased.
